dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java

```diff
@@ -41,16 +41,42 @@
 final class ConcurrencyLimitedChannel implements LimitedChannel {
     private static final Logger log = LoggerFactory.getLogger(ConcurrencyLimitedChannel.class);
 
-    private final Meter limitedMeter;
     private final NeverThrowChannel delegate;
     private final CautiousIncreaseAggressiveDecreaseConcurrencyLimiter limiter;
+    private final String channelNameForLogging;
 
-    static LimitedChannel create(Config cf, Channel channel, int uriIndex) {
+    static LimitedChannel createForHost(Config cf, Channel channel, int uriIndex) {
         ClientConfiguration.ClientQoS clientQoS = cf.clientConf().clientQoS();
         switch (clientQoS) {
             case ENABLED:
                 TaggedMetricRegistry metrics = cf.clientConf().taggedMetricRegistry();
-                return new ConcurrencyLimitedChannel(channel, createLimiter(), cf.channelName(), uriIndex, metrics);
+                CautiousIncreaseAggressiveDecreaseConcurrencyLimiter limiter = createLimiter(Behavior.HOST_LEVEL);
+                ConcurrencyLimitedChannelInstrumentation instrumentation =
+                        new HostConcurrencyLimitedChannelInstrumentation(cf.channelName(), uriIndex, limiter, metrics);
+                return new ConcurrencyLimitedChannel(channel, limiter, instrumentation);
+            case DANGEROUS_DISABLE_SYMPATHETIC_CLIENT_QOS:
+                return new ChannelToLimitedChannelAdapter(channel);
+        }
+        throw new SafeIllegalStateException(
+                "Encountered unknown client QoS configuration", SafeArg.of("ClientQoS", clientQoS));
+    }
+
+    /**
+     * Creates a concurrency limited channel for per-endpoint limiting.
+     * Metrics are not reported by this component per-endpoint, only by the per-endpoint queue.
+     */
+    static LimitedChannel createForEndpoint(
+            Channel channel,
+            String channelName,
+            int uriIndex,
+            Endpoint endpoint,
+            ClientConfiguration.ClientQoS clientQoS) {
+        switch (clientQoS) {
+            case ENABLED:
+                return new ConcurrencyLimitedChannel(
+                        channel,
+                        createLimiter(Behavior.ENDPOINT_LEVEL),
+                        new EndpointConcurrencyLimitedChannelInstrumentation(channelName, uriIndex, endpoint));
             case DANGEROUS_DISABLE_SYMPATHETIC_CLIENT_QOS:
                 return new ChannelToLimitedChannelAdapter(channel);
```

Comment:
if a user builds a client with ClientQoS disabled, ideally we want to omit the endpoint-level queue _and_ this adapter with it's concurrent hashmap right? 

I think this is implemented in DialogueChannel#L138, but doesn't that mean this branch will never be reached?

Reply:
:+1: yep, this can be simplified